File: agent_service/deepseek_client.py
import httpx
from .config import settings
from .usage_tracker import usage_tracker


class DeepSeekClient:
    def __init__(self):
        self.api_key = settings.deepseek_api_key
        self.base_url = settings.deepseek_base_url.rstrip("/")
        self.model = settings.deepseek_model

    def _headers(self) -> dict:
        return {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json",
        }

    async def chat(
        self,
        messages: list[dict],
        temperature: float = None,
        max_tokens: int = None,
    ) -> str:
        temp = temperature if temperature is not None else settings.agent_temperature
        mt = max_tokens if max_tokens is not None else settings.agent_max_tokens
        async with httpx.AsyncClient(timeout=settings.agent_api_timeout) as client:
            resp = await client.post(
                f"{self.base_url}/chat/completions",
                headers=self._headers(),
                json={
                    "model": self.model,
                    "messages": messages,
                    "temperature": temp,
                    "max_tokens": mt,
                },
            )
            resp.raise_for_status()
            data = resp.json()

            usage = data.get("usage", {})
            prompt_tokens = usage.get("prompt_tokens", 0)
            completion_tokens = usage.get("completion_tokens", 0)

            if prompt_tokens or completion_tokens:
                usage_tracker.record(
                    model=data.get("model", self.model),
                    endpoint="chat/completions",
                    prompt_tokens=prompt_tokens,
                    completion_tokens=completion_tokens,
                )

            return data["choices"][0]["message"]["content"]

    async def embed(self, texts: list[str]) -> list[list[float]]:
        if settings.embedding_provider != "deepseek":
            raise NotImplementedError("只有 deepseek embedding 目前支持")

        async with httpx.AsyncClient(timeout=60) as client:
            resp = await client.post(
                f"{self.base_url}/embeddings",
                headers=self._headers(),
                json={
                    "model": "deepseek-embed",
                    "input": texts,
                },
            )
            resp.raise_for_status()
            data = resp.json()

            usage = data.get("usage", {})
            prompt_tokens = usage.get("prompt_tokens", 0) if isinstance(usage, dict) else 0

            if prompt_tokens:
                usage_tracker.record(
                    model=data.get("model", "deepseek-embed"),
                    endpoint="embeddings",
                    prompt_tokens=prompt_tokens,
                    completion_tokens=0,
                )

            return [item["embedding"] for item in data["data"]]


client = DeepSeekClient()

File: obsidian-plugin/src/deepseek_client.ts
/**
 * DeepSeek API 客户端
 * 使用 Obsidian 的 requestUrl 进行 HTTP 调用
 */

import { requestUrl, RequestUrlResponse } from "obsidian";
import { AgentKBSettings } from "../settings";

interface ChatMessage {
  role: "system" | "user" | "assistant";
  content: string;
}

interface ChatResponse {
  choices: { message: { content: string } }[];
  usage?: {
    prompt_tokens?: number;
    completion_tokens?: number;
  };
  model?: string;
}

interface UsageRecord {
  model: string;
  endpoint: string;
  promptTokens: number;
  completionTokens: number;
  timestamp: number;
}

export class DeepSeekClient {
  private apiKey: string;
  private baseUrl: string;
  private model: string;
  private usageLog: UsageRecord[] = [];

  constructor(settings: AgentKBSettings) {
    this.apiKey = settings.deepseekApiKey;
    this.baseUrl = settings.deepseekBaseUrl.replace(/\/$/, "");
    this.model = settings.deepseekModel;
  }

  updateSettings(settings: AgentKBSettings): void {
    this.apiKey = settings.deepseekApiKey;
    this.baseUrl = settings.deepseekBaseUrl.replace(/\/$/, "");
    this.model = settings.deepseekModel;
  }

  /**
   * 调用 Chat Completions API
   */
  async chat(
    messages: ChatMessage[],
    options?: { temperature?: number; maxTokens?: number }
  ): Promise<string> {
    const temperature = options?.temperature ?? 0.3;
    const maxTokens = options?.maxTokens ?? 2048;

    const resp: RequestUrlResponse = await requestUrl({
      url: `${this.baseUrl}/chat/completions`,
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.apiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        model: this.model,
        messages,
        temperature,
        max_tokens: maxTokens,
      }),
    });

    if (resp.status !== 200) {
      throw new Error(`DeepSeek API error: ${resp.status} ${resp.text}`);
    }

    const data: ChatResponse = resp.json;
    const usage = data.usage;

    if (usage?.prompt_tokens || usage?.completion_tokens) {
      this.usageLog.push({
        model: data.model || this.model,
        endpoint: "chat/completions",
        promptTokens: usage.prompt_tokens || 0,
        completionTokens: usage.completion_tokens || 0,
        timestamp: Date.now(),
      });
    }

    return data.choices[0].message.content;
  }

  /**
   * 获取最近的用量记录
   */
  getUsage(): UsageRecord[] {
    return [...this.usageLog];
  }

  /**
   * 获取今日用量统计
   */
  getTodayUsage(): { calls: number; promptTokens: number; completionTokens: number } {
    const todayStart = new Date();
    todayStart.setHours(0, 0, 0, 0);
    const todayTs = todayStart.getTime();

    const todayRecords = this.usageLog.filter((r) => r.timestamp >= todayTs);
    return {
      calls: todayRecords.length,
      promptTokens: todayRecords.reduce((s, r) => s + r.promptTokens, 0),
      completionTokens: todayRecords.reduce((s, r) => s + r.completionTokens, 0),
    };
  }
}

// 全局单例，由 main.ts 初始化
let _client: DeepSeekClient | null = null;

export function initClient(settings: AgentKBSettings): void {
  _client = new DeepSeekClient(settings);
}

export function getClient(): DeepSeekClient {
  if (!_client) throw new Error("DeepSeekClient not initialized");
  return _client;
}
